shadow-4.0.3/man/vipw.8
.\"$Id: vipw.8 6 2005-03-20 15:34:28Z bubulle $
.TH VIPW 8 "26 Sep 1997"
.SH NAME
vipw, vigr \- edit the password, group, shadow-password, or shadow-group file.
.SH SYNOPSIS
\fBvipw\fR [\fB-s\fR]
.br
\fBvigr\fR [\fB-s\fR]
.SH DESCRIPTION
.BR vipw " and " vigr
will edit the files
.IR /etc/passwd " and " /etc/group ", respectively."
With the
.B -s
flag, they will edit the shadow versions of those files,
.IR /etc/shadow " and " /etc/gshadow ", respectively.
The programs will set the appropriate locks to prevent file corruption.

When looking for an editor, the programs will first try the
environment variable
.BR VISUAL ,
then the environment variable
.BR EDITOR ,
and finally the default editor,
.BR vi .
.SH "SEE ALSO"
.BR group (5),
.BR passwd (5),
.BR shadow (5)


noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



shadow-4.0.3/src/vipw.c
/*
  vipw, vigr  edit the password or group file
  with -s will edit shadow or gshadow file
 
  Copyright (C) 1997 Guy Maor <maor@ece.utexas.edu>

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

  */

#include <config.h>

#include "rcsid.h"
RCSID (PKG_VER "$Id: vipw.c 6 2005-03-20 15:34:28Z bubulle $")
#include "defines.h"
#include <errno.h>
#include <sys/stat.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <signal.h>
#include <utime.h>
#include "prototypes.h"
#include "pwio.h"
#include "shadowio.h"
#include "groupio.h"
#include "sgroupio.h"
static const char *progname, *filename, *fileeditname;
static int filelocked = 0, createedit = 0;
static int (*unlock) (void);

/* local function prototypes */
static int create_backup_file (FILE *, const char *, struct stat *);
static void vipwexit (const char *, int, int);
static void vipwedit (const char *, int (*)(void), int (*)(void));

static int
create_backup_file (FILE * fp, const char *backup, struct stat *sb)
{
	struct utimbuf ub;
	FILE *bkfp;
	int c;
	mode_t mask;

	mask = umask (077);
	bkfp = fopen (backup, "w");
	umask (mask);
	if (!bkfp)
		return -1;

	rewind (fp);
	while ((c = getc (fp)) != EOF) {
		if (putc (c, bkfp) == EOF)
			break;
	}

	if (c != EOF || fflush (bkfp)) {
		fclose (bkfp);
		unlink (backup);
		return -1;
	}
	if (fclose (bkfp)) {
		unlink (backup);
		return -1;
	}

	ub.actime = sb->st_atime;
	ub.modtime = sb->st_mtime;
	if (utime (backup, &ub) ||
	    chmod (backup, sb->st_mode) ||
	    chown (backup, sb->st_uid, sb->st_gid)) {
		unlink (backup);
		return -1;
	}
	return 0;
}


static void vipwexit (const char *msg, int syserr, int ret)
{
	int err = errno;

	if (filelocked)
		(*unlock) ();
	if (createedit)
		unlink (fileeditname);
	if (msg)
		fprintf (stderr, "%s: %s", progname, msg);
	if (syserr)
		fprintf (stderr, ": %s", strerror (err));
	fprintf (stderr, _("\n%s: %s is unchanged\n"), progname, filename);
	exit (ret);
}

#ifndef DEFAULT_EDITOR
#define DEFAULT_EDITOR "vi"
#endif

static void
vipwedit (const char *file, int (*file_lock) (void),
	  int (*file_unlock) (void))
{
	const char *editor;
	pid_t pid;
	struct stat st1, st2;
	int status;
	FILE *f;
	char filebackup[1024], fileedit[1024];

	snprintf (filebackup, sizeof filebackup, "%s-", file);
	snprintf (fileedit, sizeof fileedit, "%s.edit", file);
	unlock = file_unlock;
	filename = file;
	fileeditname = fileedit;

	if (access (file, F_OK))
		vipwexit (file, 1, 1);
	if (!file_lock ())
		vipwexit (_("Couldn't lock file"), errno, 5);
	filelocked = 1;

	/* edited copy has same owners, perm */
	if (stat (file, &st1))
		vipwexit (file, 1, 1);
	if (!(f = fopen (file, "r")))
		vipwexit (file, 1, 1);
	if (create_backup_file (f, fileedit, &st1))
		vipwexit (_("Couldn't make backup"), errno, 1);
	createedit = 1;

	editor = getenv ("VISUAL");
	if (!editor)
		editor = getenv ("EDITOR");
	if (!editor)
		editor = DEFAULT_EDITOR;

	if ((pid = fork ()) == -1)
		vipwexit ("fork", 1, 1);
	else if (!pid) {
#if 0
		execlp (editor, editor, fileedit, (char *) 0);
		fprintf (stderr, "%s: %s: %s\n", progname, editor,
			 strerror (errno));
		exit (1);
#else
		/* use the system() call to invoke the editor so that it accepts
		   command line args in the EDITOR and VISUAL environment vars */
		char *buf;

		buf =
		    (char *) malloc (strlen (editor) + strlen (fileedit) +
				     2);
		snprintf (buf, strlen (editor) + strlen (fileedit) + 2,
			  "%s %s", editor, fileedit);
		if (system (buf) != 0) {
			fprintf (stderr, "%s: %s: %s\n", progname, editor,
				 strerror (errno));
			exit (1);
		} else
			exit (0);
#endif
	}

	for (;;) {
		pid = waitpid (pid, &status, WUNTRACED);
		if (WIFSTOPPED (status)) {
			kill (getpid (), SIGSTOP);
			kill (getpid (), SIGCONT);
		} else
			break;
	}

	if (pid == -1 || !WIFEXITED (status) || WEXITSTATUS (status))
		vipwexit (editor, 1, 1);

	if (stat (fileedit, &st2))
		vipwexit (fileedit, 1, 1);
	if (st1.st_mtime == st2.st_mtime)
		vipwexit (0, 0, 0);

	/*
	 * XXX - here we should check fileedit for errors; if there are any,
	 * ask the user what to do (edit again, save changes anyway, or quit
	 * without saving). Use pwck or grpck to do the check.  --marekm
	 */

	createedit = 0;
	unlink (filebackup);
	link (file, filebackup);
	if (rename (fileedit, file) == -1) {
		fprintf (stderr,
			 _
			 ("%s: can't restore %s: %s (your changes are in %s)\n"),
			 progname, file, strerror (errno), fileedit);
		vipwexit (0, 0, 1);
	}

	(*file_unlock) ();
}


int main (int argc, char **argv)
{
	int flag;
	int editshadow = 0;
	char *c;
	int e = 1;
	int do_vipw;

	setlocale (LC_ALL, "");
	bindtextdomain (PACKAGE, LOCALEDIR);
	textdomain (PACKAGE);

	progname = ((c = strrchr (*argv, '/')) ? c + 1 : *argv);
	do_vipw = (strcmp (progname, "vigr") != 0);

	while ((flag = getopt (argc, argv, "ghps")) != EOF) {
		switch (flag) {
		case 'p':
			do_vipw = 1;
			break;
		case 'g':
			do_vipw = 0;
			break;
		case 's':
			editshadow = 1;
			break;
		case 'h':
			e = 0;
		default:
			printf (_("Usage:\n\
`vipw' edits /etc/passwd        `vipw -s' edits /etc/shadow\n\
`vigr' edits /etc/group         `vigr -s' edits /etc/gshadow\n\
"));
			exit (e);
		}
	}

	if (do_vipw) {
#ifdef SHADOWPWD
		if (editshadow)
			vipwedit (SHADOW_FILE, spw_lock, spw_unlock);
		else
#endif
			vipwedit (PASSWD_FILE, pw_lock, pw_unlock);
	} else {
#ifdef SHADOWGRP
		if (editshadow)
			vipwedit (SGROUP_FILE, sgr_lock, sgr_unlock);
		else
#endif
			vipwedit (GROUP_FILE, gr_lock, gr_unlock);
	}

	return 0;
}
